sdk/iot/hub/src/az_iot_hub_client_sas.c

```diff
@@ -14,104 +14,136 @@
-      + az_span_length(device_id) + az_span_length(signature) + 5;
+      + az_span_length(sig_string) + az_span_length(se_string)
+      + az_span_length(client->_internal.iot_hub_hostname)
+      + az_span_length(client->_internal.device_id) + az_span_length(base64_hmac_sha256_signature)
+      + 5; // Number of ampersands and equal signs.
 
-  AZ_RETURN_IF_NOT_ENOUGH_CAPACITY(sas_token, required_length);
+  if (az_span_length(client->_internal.options.module_id) > 0)
+  {
+    required_length
+        += az_span_length(client->_internal.options.module_id) + az_span_length(modules_string);
+  }
 
-  *out_sas_token = sas_token;
+  AZ_RETURN_IF_NOT_ENOUGH_CAPACITY(mqtt_password, required_length);
 
   // SharedAccessSignature
-  *out_sas_token = az_span_append(*out_sas_token, sr_string);
-  *out_sas_token = az_span_append_uint8(*out_sas_token, EQUAL_SIGN);
-  *out_sas_token = az_span_append(*out_sas_token, iothub_fqdn);
-  *out_sas_token = az_span_append(*out_sas_token, devices_string);
-  *out_sas_token = az_span_append(*out_sas_token, device_id);
+  mqtt_password = az_span_append(mqtt_password, sr_string);
+  mqtt_password = az_span_append_uint8(mqtt_password, EQUAL_SIGN);
+  mqtt_password = az_span_append(mqtt_password, client->_internal.iot_hub_hostname);
+
+  // Device ID
+  mqtt_password = az_span_append(mqtt_password, devices_string);
+  mqtt_password = az_span_append(mqtt_password, client->_internal.device_id);
+
+  // Module ID
+  if (az_span_length(client->_internal.options.module_id) > 0)
+  {
+    mqtt_password = az_span_append(mqtt_password, modules_string);
+    mqtt_password = az_span_append(mqtt_password, client->_internal.options.module_id);
+  }
 
   // Signature
-  *out_sas_token = az_span_append_uint8(*out_sas_token, AMPERSAND);
-  *out_sas_token = az_span_append(*out_sas_token, sig_string);
-  *out_sas_token = az_span_append_uint8(*out_sas_token, EQUAL_SIGN);
-  *out_sas_token = az_span_append(*out_sas_token, signature);
+  mqtt_password = az_span_append_uint8(mqtt_password, AMPERSAND);
+  mqtt_password = az_span_append(mqtt_password, sig_string);
+  mqtt_password = az_span_append_uint8(mqtt_password, EQUAL_SIGN);
+  mqtt_password = az_span_append(mqtt_password, base64_hmac_sha256_signature);
 
   // Expiration
-  *out_sas_token = az_span_append_uint8(*out_sas_token, AMPERSAND);
-  *out_sas_token = az_span_append(*out_sas_token, se_string);
-  *out_sas_token = az_span_append_uint8(*out_sas_token, EQUAL_SIGN);
-  AZ_RETURN_IF_FAILED(az_span_append_i32toa(*out_sas_token, expiry_time_secs, out_sas_token));
+  mqtt_password = az_span_append_uint8(mqtt_password, AMPERSAND);
+  mqtt_password = az_span_append(mqtt_password, se_string);
+  mqtt_password = az_span_append_uint8(mqtt_password, EQUAL_SIGN);
+  AZ_RETURN_IF_FAILED(
+      az_span_append_u32toa(mqtt_password, token_expiration_epoch_time, &mqtt_password));
 
   if (az_span_ptr(key_name) != NULL && az_span_length(key_name) > 0)
```

Comment:
> if (az_span_ptr(key_name) != NULL && az_span_length(key_name) > 0) [](start = 1, length = 67)

Can we check at the start of function about the capacity requirements and then do the operation?